### PR TITLE
Add utilities for common args validation

### PR DESCRIPTION
**Pull request by magicDGS**

### CODE REVIEW

File: core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+     *
+     * @param object  any {@link Object}.
+     * @param message the text message that would be passed to the exception thrown.
+     * @param <T>     object type.
+     *
+     * @return the same object (if non-null).
+     *
+     * @throws IllegalArgumentException if a {@code object == null}.
+     */
+    public static <T> T nonNull(final T object, final Supplier<String> message) {
```
Comment (lbergelson) [inline core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java]: I would add one arg versions of these that just a report a generic "this cannot be null" because in most places the message is not going to more useful than the stack trace.

File: core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
+package org.htsjdk.core.utils;
+
+import java.util.Collection;
+import java.util.function.Supplier;
+
+/**
+ * General utilities to check and validate conditions for constructors/methods.
+ */
+public final class ValidationUtils {
```
Comment (lbergelson) [inline core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java]: ParamUtils is marginally shorter which might be good since this is going to show up all over the place.

File: core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.Collection;
+import java.util.function.Supplier;
+
+/**
+ * General utilities to check and validate conditions for constructors/methods.
+ */
+public final class ValidationUtils {
+
+    // cannot be instantiated
+    private ValidationUtils() {}
```
Comment (lbergelson) [inline core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java]: 👍

File: core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+     *
+     * <p>It considers that the first index position is {@code 0}.
+     *
+     * @param index  query index.
+     * @param length collection or array size.
+     *
+     * @return same value as the input {@code index}.
+     */
+    public static int validateIndex(final int index, final int length) {
+        validateArg(index > 0, () -> "index cannot be negative: " + index);
```
Comment (lbergelson) [inline core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java]: shouldn't this be >= 0?

File: core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+     * @param message   the text message that would be passed to the exception thrown.
+     */
+    public static void validateArg(final boolean condition, final Supplier<String> message) {
+        if (!condition) {
+            throw new IllegalArgumentException(message.get());
+        }
+    }
+
+    /**
+     * Checks whether an index is within bounds.
```
Comment (lbergelson) [inline core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java]: I might reword to "Checks whether an index is within the bounds  [0, length).

File: core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+     *
+     * @param collection any Collection
+     * @param message    the text message that would be passed to the exception thrown.
+     * @param <T>        collection type.
+     *
+     * @return the same collection (if non-empty).
+     *
+     * @throws IllegalArgumentException if collection is {@code null} or empty.
+     */
+    public static <T extends Collection<?>> T nonEmpty(final T collection, final Supplier<String> message) {
```
Comment (lbergelson) [inline core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java]: I'd add a 1 param version.

File: core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    // cannot be instantiated
+    private ValidationUtils() {}
+
+    /**
+     * Checks that the condition is {@code true}.
+     *
+     * @param condition required condition.
+     * @param message   the text message that would be passed to the exception thrown.
+     */
+    public static void validateArg(final boolean condition, final Supplier<String> message) {
```
Comment (lbergelson) [inline core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java]: I would rename to just `validate`

File: core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java (hunk 2 above)
Comment (lbergelson) [inline core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java]: These should all mention what type of exception they throw.

Comment (magicDGS) [reply core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java]: Renamed.

Comment (magicDGS) [reply core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java]: Done

File: core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java (hunk 4 above)
Comment (magicDGS) [reply core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java]: Embarrasing error. Changed and added test for edge cases.

File: core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java (hunk 5 above)
Comment (magicDGS) [reply core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java]: Done.

File: core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java (hunk 7 above)
Comment (magicDGS) [reply core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java]: Done.

File: core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java (hunk 1 above)
Comment (magicDGS) [reply core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java]: Done.

File: core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java (hunk 6 above)
Comment (magicDGS) [reply core/src/main/java/org/htsjdk/core/utils/ValidationUtils.java]: Done.

**Comment by magicDGS** [comment]
Back to you @lbergelson

File: core/src/main/java/org/htsjdk/core/utils/ParamUtils.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+     *
+     * @param object  any {@link Object}.
+     * @param <T>     object type.
+     *
+     * @return the same object (if non-null).
+     *
+     * @throws IllegalArgumentException if a {@code object == null}.
+     */
+    public static <T> T nonNull(final T object) {
+        return nonNull(object, () -> object.getClass() + " cannot be null");
```
Comment (lbergelson) [inline core/src/main/java/org/htsjdk/core/utils/ParamUtils.java]: Won't dereferencing `object` here cause a null pointer?

File: core/src/main/java/org/htsjdk/core/utils/ParamUtils.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+     * Checks that a {@link Collection} is not {@code null} and that it is not empty.
+     *
+     * @param collection any Collection
+     * @param <T>        collection type.
+     *
+     * @return the same collection (if non-empty).
+     *
+     * @throws IllegalArgumentException if collection is {@code null} or empty.
+     */
+    public static <T extends Collection<?>> T nonEmpty(final T collection) {
```
Comment (lbergelson) [inline core/src/main/java/org/htsjdk/core/utils/ParamUtils.java]: I think this may throw NPE when `collection` is null here because the message gets passed into `nonNull` and then dereferences `collection`.

File: core/src/main/java/org/htsjdk/core/utils/ParamUtils.java (hunk 8 above)
Comment (magicDGS) [reply core/src/main/java/org/htsjdk/core/utils/ParamUtils.java]: Reason 1 for not coding at 23:00

File: core/src/main/java/org/htsjdk/core/utils/ParamUtils.java (hunk 9 above)
Comment (magicDGS) [reply core/src/main/java/org/htsjdk/core/utils/ParamUtils.java]: Reason 2 for not coding at 23:00

**Comment by magicDGS** [comment]
Back to you @lbergelson - and to @tfenne and @jacarey if they want to take a look!

**Comment by lbergelson** [comment]
@magicDGS Thank you :)
